### PR TITLE
run the supervisor as a service in a windows container studio

### LINKED ISSUE #4665: Windows Container Studio does not stream supervisor log
The `Get-SupervisorLog` command in a container studio simply tails the log but is not streaming the log file. This is a recent regression caused by piping the supervisor output in the ssame process as the studio shell which blocks the supervisor stream.

The right way to fix this is to Have the supervisor run as a Windows service inside the container. This will cause the supervisor to run in an entirely different console process. It also prevent `ctrl+C`s from leaking in to the Supervisor process and unintentionally killing the supervisor.

### PR DESCRIPTION
**Pull request by mwrock**
fixes #4665 

This PR addresses an issue with Windows container studios where calling `Get-SupervisorLog` does not actually stream the supervisor log because the event handler loggers are blocked on the waiting `Get-Content` which is all running in the same process. This change causes the supervisor to now run as a windows service which is isolated to its own process and command console. This has several benefits:

* The supervisor is completely isolated from stream and signal handling in the powershell process. The supervisor cannot intercept control signals from the powershell process inadvertainly killing the supervisor.
* We get some dedicated usage of the Windows-Service package which should help us to catch failures when it is updated.
* We now will need to preload the supervisor package in the docker image creation causing the supervisor to start more quickly in a studio

Signed-off-by: mwrock <matt@mattwrock.com>

### CODE REVIEW

**Comment by thesentinels** [comment]
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

**Comment by baumanj** [comment]
Whelp, time to start learning PowerShell. Do you have any favorite getting started resources @mwrock?

**Comment by mwrock** [comment]
boy its been a long time @baumanj . You might want to start here: https://github.com/PowerShell/PowerShell/tree/master/docs/learning-powershell

**Comment by baumanj** [comment]
I don't see anything that looks obviously wrong. I'm still getting up to speed on PowerShell though.

**Comment by mwrock** [comment]
To make this easier for non powershell folks to grok, I'll elaborate on exactly what the changes do:

Changes to `hab-studio.ps1` add branching logic for what to do when the supervisor is running as a service (which it will be when running a containerized studio). It includes 3 key changes:

1. When tailing the log, parse the windows service's XML config file to find where it is logging the stdout/err and tail that instead of the `out.log` in `/hab/sup/default`
2. If calling the studio's `Stop-Supervisor` function, properly stop the habitat windows service (which runs a launcher/supervisor) instead of killing the process in the `LOCK` file.
3. Only kickoff the background supervisor and setup its stream logging if the service is not installed and running. Note that almost all of these "changed" lines are just indentation since they got moved inside of an `if`.

Next the changes to `build-docker-image.ps1` actually makes sure the image gets a windows service. 

1. We add a parameter for specifying an array of pkgs to install into the image. This mimics the bash version and makes it easier to test with a studio hart file instead of the `stable` studio package.
2. We add the [`windows-service`](https://github.com/habitat-sh/windows-service) and also the supervisor which the service will need.
3. We add a `RUN` directive to the dockerfile that calls the installer that ships with our `windows-service` package. It basically installs a little C# wrapper into the Windows Service Control Manager (SCM) which controls logging and start/stop behavior for the launcher/supervisor.

**Comment by fnichol** [comment]
Ah, that's a nice benefit--getting the Supervisor a little more isolated out of the way in a Studio!

File: components/studio/bin/hab-studio.ps1 [hunk 1]
```diff
@@ -301,12 +301,22 @@ function Enter-Studio {
         Start-Process "$env:STUDIO_SCRIPT_ROOT\powershell\pwsh.exe" -ArgumentList "-Command `"& {Get-Content $env:HAB_STUDIO_ENTER_ROOT\hab\sup\default\out.log -Tail 100 -Wait}`""
       }
       else {
-        Get-Content $env:HAB_STUDIO_ENTER_ROOT\hab\sup\default\out.log -Tail 100 -Wait
```
Comment (fnichol) [inline components/studio/bin/hab-studio.ps1]: Right, this would have blocked here?

File: components/studio/bin/hab-studio.ps1 [hunk 2]
```diff
@@ -301,12 +301,22 @@ function Enter-Studio {
...
+        # Container studios run habitat as a service which logs to a 
+        # configured file location
+        $svcPath = Join-Path $env:SystemDrive "hab\svc\windows-service"
+        [xml]$configXml = Get-Content (Join-Path $svcPath HabService.exe.config)
```
Comment (fnichol) [inline components/studio/bin/hab-studio.ps1]: This is making me wonder what a `hab svc path <IDENT>` subcommand might look like, just for scripting convenience

File: components/studio/bin/hab-studio.ps1 [hunk 3]
```diff
@@ -315,28 +325,31 @@ function Enter-Studio {
...
-    
-    # As an industry, this should make us all feel bad
-    # without this stdin never seems to come back and the PS prompt
-    # fails to appear and things look like they are hung
-    Start-Sleep -Milliseconds 100
-
-    $pr.BeginOutputReadLine()
+    $habSvc = Get-Service Habitat -ErrorAction SilentlyContinue
+    if(!$habSvc -or ($habSvc.Status -eq "Stopped")) {
+      $pr = New-Object System.Diagnostics.Process
```
Comment (fnichol) [inline components/studio/bin/hab-studio.ps1]: One of my favorite things about PowerShell, the .net stdlib is right there!

File: components/studio/bin/hab-studio.ps1 [hunk 4]
```diff
@@ -315,28 +325,31 @@ function Enter-Studio {
...
+      $pr.StartInfo.UseShellExecute = $false
+      $pr.StartInfo.CreateNoWindow = $true
+      $pr.StartInfo.RedirectStandardOutput = $true
+      $pr.StartInfo.RedirectStandardError = $true
+      $pr.StartInfo.FileName = "hab.exe"
+      $pr.StartInfo.Arguments = "sup run"
+      Register-ObjectEvent -InputObject $pr -EventName OutputDataReceived -action {
+        $Event.SourceEventArgs.Data | Out-File $env:HAB_STUDIO_ENTER_ROOT\hab\sup\default\out.log -Append -Encoding ascii
+      } | Out-Null
```
Comment (fnichol) [inline components/studio/bin/hab-studio.ps1]: That's pretty cool IO redirection--this way it logs to `out.log` but doesn't also log to `stdout`?

File: components/studio/bin/hab-studio.ps1 [hunk 5]
```diff
@@ -315,28 +325,31 @@ function Enter-Studio {
...
+      Register-ObjectEvent -InputObject $pr -EventName ErrorDataReceived -action {
+        $Event.SourceEventArgs.Data | Out-File $env:HAB_STUDIO_ENTER_ROOT\hab\sup\default\out.log -Append -Encoding ascii
+      } | Out-Null
+      $pr.start() | Out-Null
+      $pr.BeginErrorReadLine()
+      
+      # As an industry, this should make us all feel bad
+      # without this stdin never seems to come back and the PS prompt
+      # fails to appear and things look like they are hung
+      Start-Sleep -Milliseconds 100
```
Comment (fnichol) [inline components/studio/bin/hab-studio.ps1]: I remember this one…

File: components/studio/bin/hab-studio.ps1 (hunk 1 above)
Comment (mwrock) [reply components/studio/bin/hab-studio.ps1]: yeah which is actually the right thing to do. Its just like a `tail -f` in bash requiring a `ctrl+c` to quit.

File: components/studio/bin/hab-studio.ps1 (hunk 2 above)
Comment (mwrock) [reply components/studio/bin/hab-studio.ps1]: yeah fur sure

File: components/studio/bin/hab-studio.ps1 (hunk 4 above)
Comment (mwrock) [reply components/studio/bin/hab-studio.ps1]: yup `out-null` == `/dev/null`

**Comment by mwrock** [comment]
@thesentinels approve

**Comment by thesentinels** [comment]
:metal: I am testing your branch against master before merging it. We do this to ensure that the master branch is never failing tests.

**Comment by thesentinels** [comment]
:neckbeard: Travis CI has [started testing this PR](https://travis-ci.org/habitat-sh/habitat/builds/347485920).

**Comment by baumanj** [comment]
> Note that almost all of these "changed" lines are just indentation since they got moved inside of an `if`.

Did you know you can view [the diff with whitespace changes suppressed](https://github.com/habitat-sh/habitat/pull/4667/files?w=1) by appending `?w=1`? If I have a PR with a lot of formatting changes, I'll sometimes include that link in the description.

**Comment by mwrock** [comment]
Oh yes that is very helpful!

**Pull request merged**